### DISCUSSION TITLE
请问如何能得知数据已经flush

### DISCUSSION THREAD

**Discussion by herowzz** [apache/iotdb#3839] [Q&A] [answered]
通过程序往DB里插入数据后，有的时候data文件夹里内容为空，需手工执行flush命令后才能刷新过去
请问如果我在执行定时备份的时候，想知道数据是否已经flush，请问有什么办法能够得知数据已刷新过去吗？

**Comment by qiaojialin** [comment, marked as answer]
可以检查 data/wal 目录里的文件创建时间，这些时间之前的数据都已经持久化到 TsFile 中了。
如果是定时备份，最好备份之前执行一下 flush。

背景知识：
数据先写到内存中的 memtable，同时会记录 wal，wal 是每个 TsFile 的 memtable 会对应一个，memtable 持久化之后对应的 wal 会被删除。
手动执行 flush 命令可以将系统当前所有 memtable 持久化+删掉 wal+将TsFile封口。但是如果还有写入，还会产生 wal。

**Reply by herowzz** [reply]
如果通过程序执行flush指令的话，例如：session.executeNonQueryStatement("flush")
调用之后是同步返回结果还是异步返回结果？这行代码执行完毕之后，flush就算完成了吗？

**Reply by qiaojialin** [reply]
同步的

**Comment by myysophia** [comment]
1.1.2 iotdb, 恢复后datanode有如下日志:
`2024-05-03 06:54:11,273 [pool-15-IoTDB-Compaction_Schedule-root.emsplus-0-1] INFO  o.a.i.d.e.c.s.i.RewriteCrossSpaceCompactionSelector:288 - root.emsplus-0 [Compaction] Total source files: 15 seqFiles, 6 unseqFiles. Candidate source files: 15 seqFiles, 6 unseqFiles. Cannot select any files because they do not meet the conditions or may be occupied by other compaction threads.
2024-05-03 06:54:21,521 [pool-29-IoTDB-LogDispatcher-DataRegion[0]-1] INFO  o.a.i.d.w.n.WALNode$PlanNodeIterator:662 - timeout when waiting for next WAL entry ready, execute rollWALFile. Current search index in wal buffer is 13733221, and next target index is 1`. 
这会导致数据不一致吗？
